src/Interpreters/ArithmeticOperationsInAgrFuncOptimize.cpp

```diff
@@ -0,0 +1,313 @@
+                        ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+    }
+
+    return {const_num, not_const};
+}
+
+std::pair<ASTs, ASTs> findAllConsts(ASTFunction * func_node, const char * inter_func_name)
+{
+    if (!func_node->arguments)
+        return {};
+
+    if (onlyConstsInside(func_node))
+        return tryGetConst(func_node->name.c_str(), func_node->arguments->children);
+    else if (inappropriateNameInside(func_node, inter_func_name))
+    {
+        if (func_node->arguments->children[0]->as<ASTLiteral>() &&
+            isConstantField(func_node->arguments->children[0]->as<ASTLiteral>()->value))
+            return {{func_node->arguments->children[0]}, {func_node->arguments->children[1]}};
+        else if (func_node->arguments->children.size() == 2 &&
+                 func_node->arguments->children[1]->as<ASTLiteral>() &&
+                 isConstantField(func_node->arguments->children[1]->as<ASTLiteral>()->value))
+            return {{func_node->arguments->children[1]}, {func_node->arguments->children[0]}};
+        else
+        {
+            if (isInappropriate(func_node->arguments->children[0], inter_func_name) && isInappropriate(func_node->arguments->children[1], inter_func_name))
+                return {{}, {func_node->arguments->children[0], func_node->arguments->children[1]}};
+            else if (isInappropriate(func_node->arguments->children[0], inter_func_name))
+            {
+                std::pair<ASTs, ASTs> ans = findAllConsts(func_node->arguments->children[1]->as<ASTFunction>(), inter_func_name);
+                ans.second.push_back(func_node->arguments->children[0]);
+                return ans;
+            }
+            else
+            {
+                std::pair<ASTs, ASTs> ans = findAllConsts(func_node->arguments->children[0]->as<ASTFunction>(), inter_func_name);
+                ans.second.push_back(func_node->arguments->children[1]);
+                return ans;
+            }
+        }
+    }
+    else
+    {
+        std::pair<ASTs, ASTs> fl = tryGetConst(func_node->name.c_str(), func_node->arguments->children);
+        ASTs first_lvl_consts = fl.first;
+        ASTs first_lvl_not_consts = fl.second;
+        if (!first_lvl_not_consts[0]->as<ASTFunction>())
+            return {first_lvl_consts, first_lvl_not_consts};
+
+        std::pair<ASTs, ASTs> ans = findAllConsts(first_lvl_not_consts[0]->as<ASTFunction>(), inter_func_name);
+        ASTs all_consts = ans.first;
+        ASTs all_not_consts = ans.second;
+
+        if (first_lvl_consts.size() == 1)
+        {
+            if (!first_lvl_not_consts[0]->as<ASTFunction>())
+                all_not_consts.push_back(first_lvl_not_consts[0]);
+
+            all_consts.push_back(first_lvl_consts[0]);
+        }
+        else if (first_lvl_consts.empty())
+        {
+            if (!first_lvl_not_consts[0]->as<ASTFunction>() || first_lvl_not_consts[0]->as<ASTFunction>()->name != inter_func_name)
```

Comment:
What's the difference in this condition and result of `isInappropriate()`? If none use function here.